### PR TITLE
Support OpenAPI v3

### LINKED ISSUE #1578: [master] Completion / tooltip does not work on yaml tab
**What happened**:
Completion / tooltip does not work on `Shoot` yaml tab 

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know**:

**Environment**:

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Alternative Implementation for the OpenAPI v3 specification.
With this approach we remove the `allOf, anyOf, oneOf` level by adding all properties directly to the parent.
This way, we get an OpenAPI v2 compatible representation and can keep our current implementation.
This also supports multiple entries in the allOf arrays as well as mixing with anyOf or oneOf 

**Which issue(s) this PR fixes**:
Fixes #1578

**Special notes for your reviewer**:
Alternative for #1594

**Release note**:

```bugfix user
Fixed code completion and tooltips in cluster editor: Recent Gardener releases dropped support for OpenAPI v2. Dashboard now uses OpenAPI v3 to fetch shoot resource information
```

### CODE REVIEW

**Comment by petersutter** [comment]
there is a build error
```
#17 ERROR: process "/bin/sh -c yarn workspace @gardener-dashboard/kube-client run lint" did not complete successfully: exit code: 1
------
 > [builder 12/16] RUN yarn workspace @gardener-dashboard/kube-client run lint:
4.410 
4.410 /app/packages/kube-client/__tests__/non-resource-endpoints.test.js
4.410   77:137  error  Parsing error: Unexpected token get
4.410 
4.410 ✖ 1 problem (1 error, 0 warnings)
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 1]
```diff
@@ -316,4 +329,39 @@ export class ShootEditorCompletions {
...
+        } else {
+          this.logger.warn('Unsupported schema array length, %s has length %i at %s', propertyName, schema[propertyName].length, currentItem)
+        }
+
+        this._resolveSchemaArrays(properties, currentItem, false)
+        Object.assign(schema, properties)
+        delete schema[propertyName]
+      } else if (typeof schema[propertyName] === 'object') {
+        currentItem = propertyName
+        this._resolveSchemaArrays(schema[propertyName], currentItem, false)
```
Comment (petersutter) [inline frontend/src/utils/shootEditorCompletions.js:362]:
```suggestion
        this._resolveSchemaArrays(properties, currentItem)
        Object.assign(schema, properties)
        delete schema[propertyName]
      } else if (typeof schema[propertyName] === 'object') {
        currentItem = propertyName
        this._resolveSchemaArrays(schema[propertyName], currentItem)
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 2]
```diff
@@ -316,4 +329,39 @@ export class ShootEditorCompletions {
 
     return { lineTokens, lineString }
   }
+
+  _resolveSchemaArrays (schema, currentItem = '') {
+    for (const propertyName in schema) {
+      let properties
+      if (['allOf', 'anyOf', 'oneOf'].includes(propertyName) && !schema[propertyName].type) {
+        if (properties) {
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:337]: This is never true

File: frontend/src/utils/shootEditorCompletions.js [hunk 3]
```diff
@@ -27,6 +29,9 @@ import {
 
 export class ShootEditorCompletions {
   constructor (shootProperties, editorIndent, supportedPaths) {
+    this.logger = useLogger()
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:32]: The Logger should be passed as paramter and defaulted to `useLogger`

File: frontend/src/utils/shootEditorCompletions.js [hunk 4]
```diff
@@ -122,17 +127,25 @@ export class ShootEditorCompletions {
       return `${propertyName}: `
     }
 
+    const generateTypeText = (type, format) => {
+      return format ? `${upperFirst(type)} (${format})` : upperFirst(type)
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:131]:
```suggestion
      if (Array.isArray(type)) {
        return type
          .map((type, i) => generateTypeText(type, format[i]))
          .join(' | ')
      }
      return format 
        ? `${upperFirst(type)} (${format})` 
        : upperFirst(type)
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 5]
```diff
@@ -122,17 +127,25 @@ export class ShootEditorCompletions {
...
+    }
+
     forIn(completions, (completion, propertyName) => {
       let text = generateCompletionText(propertyName, completion.type, token.type)
       if (token.type === 'firstArrayItem') {
         text = '- ' + text
       }
       const string = propertyName.toLowerCase()
+      const type = Array.isArray(completion.type)
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:140]:
```suggestion
      const type = generateTypeText(completion.type, completion.format)
```

File: frontend/src/utils/shootEditorCompletions.js [hunk 6]
```diff
@@ -316,4 +329,39 @@ export class ShootEditorCompletions {
 
     return { lineTokens, lineString }
   }
+
+  _resolveSchemaArrays (schema, currentItem = '') {
+    for (const propertyName in schema) {
+      let properties
+      if (['allOf', 'anyOf', 'oneOf'].includes(propertyName) && !schema[propertyName].type) {
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:336]: log an error if we have a `type` property in this case

File: frontend/src/utils/shootEditorCompletions.js [hunk 7]
```diff
@@ -316,4 +329,39 @@ export class ShootEditorCompletions {
...
+        if (properties) {
+          // We currently do not support merging when schema has multiple discriminators on same level
+          this.logger.warn('Found multiple discriminators in schema at %s', currentItem)
+        }
+        if (schema[propertyName].length === 1) {
+          properties = schema[propertyName][0]
+        } else if (schema[propertyName].length > 1 &&
+            propertyName === 'oneOf' &&
+            !schema.type &&
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:345]: I don't understand why we check for parent.type here?

File: frontend/src/utils/shootEditorCompletions.js [hunk 8]
```diff
@@ -316,4 +329,39 @@ export class ShootEditorCompletions {
...
+            schema[propertyName].every(item => {
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:346]: Extract into an own helper function

File: frontend/src/utils/shootEditorCompletions.js [hunk 9]
```diff
@@ -316,4 +329,39 @@ export class ShootEditorCompletions {
...
+              return (Object.getOwnPropertyNames(item).length === 1 && !!item.type) ||
+                (Object.getOwnPropertyNames(item).length === 2 && !!item.type && !!item.format)
```
Comment (holgerkoser) [inline frontend/src/utils/shootEditorCompletions.js:348]: Do not use !! here

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**